### PR TITLE
#9625 - Refactor: Variables are missing in props validation (let's rewrite JS to TS) 43

### LINKED ISSUE #9625: Refactor: Variables are missing in props validation (let's rewrite JS to TS) 43
**Problem:** Variables are missing in props validation
React components should validate prop types

**Why is this an issue?**
In JavaScript, props are typically passed as plain objects, which can lead to errors and confusion when working with components that have specific prop requirements. However, it lacks of type safety and clarity when passing props to components in a codebase.

By defining types for component props, developers can enforce type safety and provide clear documentation for the expected props of a component. This helps catch potential errors at compile-time. It also improves code maintainability by making it easier to understand how components should be used and what props they accept.

**How can I fix it?**
Let re-write the entire file on TypeScript and define types for all props.
Please check if types you you want to declare already exist and if no - create new declatations at types.ts if they can be used elsewhere (if yes - use existed one)

**Problem locations:**
packages/ketcher-core/src/application/render/restruct/reloop.js

**IMPORTANT:** Please do not delete old jsx file and create new tsx one, but add changes to existed file and rename it. That let me easily compare changes you made

### PR DESCRIPTION
**Pull request by mariam-khutuashvili**
Summary

  Converts packages/ketcher-core/src/application/render/restruct/reloop.js to TypeScript with full prop/parameter type annotations.
  Pure type refactor — no logic changes.

  Changes

  Single file: packages/ketcher-core/src/application/render/restruct/reloop.js → reloop.ts (renamed via git mv to preserve history).

  - ReLoop class fields typed: loop: Loop, centre: Vec2, radius: number
  - All method signatures typed: constructor(loop: Loop), static isSelectable(): boolean, show(restruct: ReStruct, _rlid: number, options: RenderOptions): void, isValid(struct: Struct, rlid: number): boolean
  - ReStruct and RenderOptions imported as import type (used only at type positions, matches the convention used by all 17 sibling files in restruct/)
  - Added a small getFromMap<T>(map, key): T helper that throws on missing keys — replaces 11 implicit non-null Map lookups with typed access and clear error messages
  - Fixed dormant JS bug: this.radius = new Vec2() → this.radius = 0 (radius is always treated as a number; the initial Vec2 was
  overwritten before being read)

  Out of scope

  - Tests — no other files modified per task scope
  - restruct.ts non-null assertion (new ReLoop(...!)) — separate file, not in scope of this PR
  - The pre-existing minSin === 0 → NaN latent edge case — out of scope for a pure JS→TS conversion (Copilot flagged this as low-confidence in a previous round; the same logic exists in master's reloop.js)

  Verification

  - npm run test — passes (prettier + eslint + types + jest for 4 packages)
  - npm run test:types — passes across all workspaces
  - Full Playwright autotest suite (Docker) — passes

## Check list
- [ ] unit-tests written
- [ ] e2e-tests written
- [ ] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [ ] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"
- [ ] reviewers are notified about the pull request

### CODE REVIEW

**Comment by AlexeyGirin** [comment]
Look in the editor

File: packages/ketcher-core/src/application/render/restruct/reloop.ts [hunk 1]
```diff
@@ -14,64 +14,87 @@
...
+import type ReStruct from './restruct';
+import type { RenderOptions } from '../render.types';
+
+function getFromMap<T>(map: Map<number, T>, key: number): T {
+  const value = map.get(key);
+  if (value === undefined) {
+    throw new Error(`Map entry not found for key ${key}`);
+  }
+  return value;
+}
```
[the remark this answers is not in the document]
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: dropped the helper and switched all the Map.get() calls in this file to ! to match the convention used in rebond.ts, reatom.ts, and restruct.ts. 
 Tried the suggestion in 0b82209 but had to revert in 84b9911. CI failed with 12 @typescript-eslint/no-non-null-assertion errors in this file. Turns out the rule is "error" package-wide at the ketcher-core root and .eslintrc.json has a per-file override list (~50 entries: reatom.ts, rebond.ts, restruct.ts etc.) that downgrades it to "warn" for those legacy files — reloop.ts is not in that list, so ! here breaks CI even though the sibling files compile. The visible "convention" in restruct/ is actually grandfathered legacy, not current policy. Keeping getFromMap for now as the lint-compliant equivalent.

File: packages/ketcher-core/src/application/render/restruct/reloop.ts [hunk 2]
```diff
@@ -14,64 +14,87 @@
...
 class ReLoop extends ReObject {
-  constructor(loop) {
+  loop: Loop;
+  centre: Vec2;
+  radius: number;
+
+  constructor(loop: Loop | undefined) {
     super('loop');
+    assert(loop !== undefined, 'ReLoop requires a Loop instance');
     this.loop = loop;
```
[the remark this answers is not in the document]
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Agreed — the constructor now takes Loop directly, the runtime assert is gone, and the single caller in restruct.ts:627 uses this.molecule.loops.get(loopId)! (which is allowed there since restruct.ts is in the per-file "warn" override list).

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexeyGirin) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Drop function and do not use Clause suggestion but use check on undefined in every place

File: packages/ketcher-core/src/application/render/restruct/restruct.ts [hunk 3]
```diff
@@ -624,7 +624,7 @@ class ReStruct {
       this.markBond(bid, 1);
     });
     ret.newLoops.forEach((loopId) => {
-      this.reloops.set(loopId, new ReLoop(this.molecule.loops.get(loopId)));
+      this.reloops.set(loopId, new ReLoop(this.molecule.loops.get(loopId)!));
```
Comment (AlexeyGirin) [inline packages/ketcher-core/src/application/render/restruct/restruct.ts]: Use ===undefined check and call return if so

**Review by AlexeyGirin — DISMISSED** [review]

File: packages/ketcher-core/src/application/render/restruct/restruct.ts (hunk 3 above)
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/restruct.ts]: Done in e7c33440:

  ret.newLoops.forEach((loopId) => {
    const loop = this.molecule.loops.get(loopId);
    if (loop === undefined) return;
    this.reloops.set(loopId, new ReLoop(loop));
  });

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Removed getFromMap and replaced every Map.get() call site with an explicit if (… === undefined) return; check. Converted the three forEach loops to for…of so the early return aborts the whole show method (matching the original throw-on-missing semantics) instead of just skipping one iteration.

File: packages/ketcher-core/src/application/render/restruct/reloop.ts [hunk 4]
```diff
@@ -16,95 +16,100 @@
...
 
-  static isSelectable() {
+  static isSelectable(): boolean {
     return false;
   }
 
-  show(restruct, rlid, options) {
-    // eslint-disable-line max-statements
-    const render = restruct.render;
+  // eslint-disable-next-line max-statements
```
Comment (svvald) [inline packages/ketcher-core/src/application/render/restruct/reloop.ts]: I know it was already present but while we're here maybe we could get rid of this eslint bypass?

File: packages/ketcher-core/src/application/render/restruct/reloop.ts [hunk 5]
```diff
@@ -16,95 +16,100 @@
...
-        options,
-      );
+      if (bond === undefined) return;
+      const beginAtom = restruct.atoms.get(hb.begin);
+      if (beginAtom === undefined) return;
+      const apos = Scale.modelToCanvas(beginAtom.a.pp, options);
       if (bond.b.type !== Bond.PATTERN.TYPE.AROMATIC) loop.aromatic = false;
-      this.centre.add_(apos); // eslint-disable-line no-underscore-dangle
-    });
+      // eslint-disable-next-line no-underscore-dangle
```
Comment (svvald) [inline packages/ketcher-core/src/application/render/restruct/reloop.ts]: And this one as well. I wonder why `add_` was used here, why not `add`?

File: packages/ketcher-core/src/application/render/restruct/reloop.ts [hunk 6]
```diff
@@ -16,95 +16,100 @@
...
-    const molecule = restruct.molecule;
-    const loop = this.loop;
+    const { loop } = this;
+    const { hbs } = loop;
+
     this.centre = new Vec2();
-    loop.hbs.forEach((hbid) => {
+    for (const hbid of hbs) {
       const hb = molecule.halfBonds.get(hbid);
+      if (hb === undefined) return;
```
Comment (svvald) [inline packages/ketcher-core/src/application/render/restruct/reloop.ts]: A nitpick on this and all other `if` oneliners here, let's stick to having all of them in newline:
```
if (...) {
return; // or any other action e.g. loop.aromatic = false
}
```

It will drastically improve the code clarity and readability

File: packages/ketcher-core/src/application/render/restruct/reloop.ts [hunk 7]
```diff
@@ -16,95 +16,100 @@
...
+    const { render, molecule } = restruct;
     const paper = render.paper;
-    const molecule = restruct.molecule;
-    const loop = this.loop;
+    const { loop } = this;
+    const { hbs } = loop;
+
     this.centre = new Vec2();
-    loop.hbs.forEach((hbid) => {
+    for (const hbid of hbs) {
```
Comment (svvald) [inline packages/ketcher-core/src/application/render/restruct/reloop.ts]: Also while we're here I'd suggest to refactor local variables names and make them more verbose for better clarity (hbs -> halfBonds or halfBondsIds, hbid -> halfBondId etc.)

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 6 above)
Comment (svvald) [inline packages/ketcher-core/src/application/render/restruct/reloop.ts]: Also I think this (and most or all the check below) could be simplified as `(!hb)` as `halfBonds` pool returns either `HalfBond` entity or `undefined` and the former could not be falsy but meaningful (in contrast with entity id which can be 0 which is falsy but meaningful)

**Review by svvald — CHANGES_REQUESTED** [review]

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 4 above)
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Turns out max-statements isn't configured anywhere in our eslint chain (not in `standard`, `@typescript-eslint/recommended`, or the package rules), so the directive was a no-op — eslint passes clean with it removed, no extraction needed.

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 5 above)
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Switched to `this.centre = this.centre.add(beginPosition)` — Vec2.add returns a new vector, so it's equivalent, reads clearer, and the no-underscore-dangle bypass is gone.

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 6 above)
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Done — every `if` body is now on its own line with braces.

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 7 above)
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Done — renamed throughout: hbs→halfBondIds, hbid→halfBondId, hb→halfBond, hba/hbb→currentHalfBond/nextHalfBond, k→index, apos/bpos→beginPosition/endPosition, n→normal, dist→distance, pi/qi→atomPosition/innerPosition, dir→direction, pathStr→pathString.

File: packages/ketcher-core/src/application/render/restruct/reloop.ts (hunk 6 above)
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/application/render/restruct/reloop.ts]: Done — all the half-bond / bond / atom entity checks now use `!entity`. I kept `firstAtomId === undefined` as an explicit check, since that one's an atom id and 0 is falsy-but-meaningful — exactly the distinction you called out.

**Review by svvald — APPROVED** [review]
